src/DataStructures/CompressedMatrix.hpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// `blaze::CompressedMatrix` is a general-purpose sparse matrix type. This file
+/// implements interoperability of `blaze::CompressedMatrix` with our data
+/// structures.
+
+#pragma once
+
+#include <blaze/math/CompressedMatrix.h>
+#include <cstddef>
+#include <pup.h>
+#include <vector>
+
+#include "Options/ParseOptions.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace PUP {
+/// @{
+/// Serialization of blaze::CompressedMatrix
+template <typename Type, bool SO, typename Tag>
+void pup(er& p, blaze::CompressedMatrix<Type, SO, Tag>& t) {
+  size_t rows = t.rows();
+  size_t columns = t.columns();
+  p | rows;
+  p | columns;
+  const size_t first_dimension = (SO == blaze::rowMajor) ? rows : columns;
+  size_t num_non_zeros;
+  size_t index;
+  if (p.isUnpacking()) {
+    t.resize(rows, columns);
+    for (size_t i = 0; i < first_dimension; ++i) {
+      p | num_non_zeros;
+      for (size_t j = 0; j < num_non_zeros; ++j) {
+        p | index;
+        if constexpr (SO == blaze::rowMajor) {
```

Comment:
This is a surprising implementation to me. Naively I expected something like `p | t.data();`. Charm++ used to have serialization of `std::vector` be done element-by-element and it was terribly slow. If there's a way to do this across a contiguous data set with `PUP_array_bytes` then it'll be a lot faster, except if you have a sparse matrix with way more data storage than data, but presumably those cases should be avoided in general :) If there isn't a way to do a contiguous stream, could you please add a brief comment why not so that in the future if someone looks at this implementation they'll at least know it's intentional?

Reply:
Pushed a fixup that switches the serialization to use Blaze's low-level `append` mechanism. We can't use `PUParray` because the sparse types don't expose `.data()`, since their storage is a combination of values and indices. I added a comment that there may be a faster way to do this copying raw memory, but I haven't figured it out yet.